File: Java Solutions/MergeTwoSortedLists.java
/**
  * Solution description
  * Iterate through both lists, check to see which node is lowest.
  * Add the lowest node to the list. Continue iterating until one list reaches the end
  * Once one list ends, just add the remainder of the leftover list to the new list
  */

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode headNode = new ListNode();
        ListNode currentNode = headNode;

        while (list1 != null && list2 != null)
        {
            // Iterate through both lists and add the lowest value to the merged list
            // Do this until one of the lists run out
            if (list1.val < list2.val)
            {
                currentNode.val = list1.val;
                list1 = list1.next;
            }

            else
            {
                currentNode.val = list2.val;
                list2 = list2.next;
            }
            
            currentNode.next = new ListNode();
            currentNode = currentNode.next;
        }

        if (list1 != null)
        {
            // List 1 still has leftover values
            currentNode.val = list1.val;
            currentNode.next = list1.next;
        }
        else if (list2 != null)
        {
            // List 2 still has leftover values
            currentNode.val = list2.val;
            currentNode.next = list2.next;
        }
        else
        {
            // The only way for list 1 and list 2 to both be null is for them to both be empty.
            // In this case, returning null as the headNode is the desired action
            headNode = null;
        }

        return headNode;
    }
}

File: C# Solutions/MergeTwoSortedLists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

/**
 * Solution description:
 *
 * Iterate through both lists at the same time, if both values are the same, put both values into the new list.
 * If the two values are different, place the lowest value in first and increment your position on that list.
 * Continue to inch along each array until they're both merged into one list
 */

public class Solution {
    public ListNode MergeTwoLists(ListNode list1, ListNode list2) {

        // HeadNode
        ListNode HeadMergedNode = new ListNode();
        ListNode CurrentNode = HeadMergedNode;
        
        while (list1.next != null && list2.next != null)  // Iterate until you reach the end of one of the lists
        {
            //Find lowest value and add it to the merged list
            
            if (list1.val < list2.val)
            {
                // Add list1 val into merged list
                // Move to next node on list1
                CurrentNode.val = list1.val;
                list1 = list1.next;
            }

            else
            {
                // Add list2 val into merged list
                // Move to next node on list2
                CurrentNode.val = list2.val;
                list2 = list2.next;
            }

            // Create new node in merged list
            CurrentNode.next = new ListNode();
            CurrentNode = CurrentNode.next; 
        }
        //return HeadMergedNode; 
        return CurrentNode;
    }
}